src/Modules/Visualization/ShowField.cc

```diff
@@ -445,757 +486,327 @@ void GeometryBuilder::renderFacesLinear(
   mesh->synchronize(Mesh::FACES_E);
 
   VMesh::Face::size_type numFaces;
-  int numAttributes = 3; //intially 3 because we will atleast be rendering verticies (vec3's)
-
   mesh->size(numFaces);
-  if (numFaces == 0)
-    return;
+  if (numFaces == 0) return;
 
-  bool withNormals = state.get(RenderState::USE_NORMALS);
-  if (withNormals)
+  Point idpt;
+  VMesh::Face::iterator fiter, fiterEnd;
+  VMesh::Node::array_type nodes;
+  mesh->get_nodes(nodes, *fiter);
+  mesh->get_point(idpt, nodes[0]);
+  mesh->begin(fiter);
+  mesh->end(fiterEnd);
+  int numNodesPerFace = nodes.size();
+  bool useQuads = (numNodesPerFace == 4);
+  int numAttributes = 3; //intially 3 because we will atleast be rendering verticies (vec3's)
+
+  bool useNormals = state.get(RenderState::USE_NORMALS);
+  bool useFaceNormals = state.get(RenderState::USE_FACE_NORMALS) && mesh->has_normals();
+  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  if (useNormals)
   {
     numAttributes += 3;
     mesh->synchronize(Mesh::NORMALS_E);
   }
 
-  bool invertNormals = state_->getValue(ShowField::FaceInvertNormals).toBool();
+  bool useColorMap = (fld->basis_order() >= 0 && state.get(RenderState::USE_COLORMAP));
+  bool isCellData = (fld->basis_order() == 0 && mesh->dimensionality() == 3);
+  bool isFaceData = (fld->basis_order() == 0 && mesh->dimensionality() == 2);
+  bool isNodeData = (fld->basis_order() == 1);
+  bool isScalar = fld->is_scalar();
+  bool isVector = fld->is_vector();
+  bool isTensor = fld->is_tensor();
+  int colorMapCase = (isCellData * 0 + isFaceData * 1 + isNodeData * 2) * 3;
+  colorMapCase += isScalar * 0 + isVector * 1 + isTensor * 2;
+
   ColorScheme colorScheme = ColorScheme::COLOR_UNIFORM;
-  std::vector<double> svals;
-  std::vector<Vector> vvals;
-  std::vector<Tensor> tvals;
-  std::vector<ColorRGB> face_colors;
 
-  if (fld->basis_order() < 0 || state.get(RenderState::USE_DEFAULT_COLOR))
-  {
-    colorScheme = ColorScheme::COLOR_UNIFORM;
-  }
-  else if (state.get(RenderState::USE_COLORMAP))
+  auto realColorMap = colorMap.get();
```

Comment:
There's something just not right with this block (524-532). It should be rewritten into one function and tested separately.

Reply:
It's leading to the crash with grayscale coloring.

Reply:
If I set a default color in ShowField it does not crash

Reply:
Right, move the creation of `map` to the block where it is used, and rename it. That should fix the crash on 10.10. Then, be more explicit about the need to create a default colormap just to create a coordinateMap--can you just compute the scale and shift?. Plus the constant parameters for the coordinateMap are magic numbers. I'd like to see a function call `createCoordinateMap` that encapsulates all that.

Reply:
@Haydelj required review changes above